internal/contentmanager/source/dynamicsource.go

```diff
@@ -0,0 +1,223 @@
+	return fmt.Sprintf("contentmanager.DynamicInformerSource - GVR: %s, Owner Type: %T, Owner Name: %s", dis.cfg.GVR, dis.cfg.Owner, dis.cfg.Owner.GetName())
+}
+
+func (dis *dynamicInformerSource) Start(ctx context.Context, q workqueue.TypedRateLimitingInterface[reconcile.Request]) error {
+	// Lock to check if this source has already been started.
+	// This should be the only time we need to lock, even if this
+	// function is called concurrently. This is because the first
+	// call that is able to lock and set the started value to
+	// true would "win" and be the one that continues on.
+	dis.mu.Lock()
+	if dis.started {
+		dis.mu.Unlock()
+		return errors.New("already started")
+	}
+
+	dis.started = true
+	dis.mu.Unlock()
+
+	informerCtx, cancel := context.WithCancel(ctx)
+
+	gInf := dis.cfg.DynamicInformerFactory.ForResource(dis.cfg.GVR)
+	dis.errChan = make(chan error, 1)
+	dis.syncedChan = make(chan struct{}, 1)
+	dis.informer = gInf
+	dis.cancel = cancel
+	dis.ctx = informerCtx
+	eventHandler := source.NewEventHandler(ctx, q, dis.cfg.Handler, dis.cfg.Predicates)
+
+	// If we encounter an error during the watch we will:
+	// - Capture the error
+	// - cancel the informer
+	// requeuing of the ClusterExtension should happen by the
+	// WaitForSync function returning an error
+	// Only if we have successfully synced in the past should we
+	// requeue the ClusterExtension
+	sharedIndexInf := gInf.Informer()
+	err := sharedIndexInf.SetWatchErrorHandler(func(r *cgocache.Reflector, err error) {
+		dis.mu.Lock()
+		dis.errOnce.Do(func() {
+			dis.errChan <- err
+			dis.err = err
+			close(dis.errChan)
+		})
+		dis.cancel()
+
+		if dis.synced {
+			// We won't be able to update the ClusterExtension status
+			// conditions so instead force a requeue if we
+			// have previously synced and then errored
+			q.Add(reconcile.Request{
+				NamespacedName: types.NamespacedName{
+					Name: dis.cfg.Owner.GetName(),
+				},
+			})
+			dis.cfg.OnPostSyncError(dis.ctx)
+		}
+		dis.mu.Unlock()
+		cgocache.DefaultWatchErrorHandler(r, err)
+	})
+	if err != nil {
+		return fmt.Errorf("setting WatchErrorHandler: %w", err)
+	}
+
+	_, err = sharedIndexInf.AddEventHandler(eventHandler.HandlerFuncs())
+	if err != nil {
+		return fmt.Errorf("adding event handler: %w", err)
+	}
+
+	go sharedIndexInf.Run(informerCtx.Done())
+
+	go func() {
+		syncOnce := sync.OnceFunc(func() {
+			dis.mu.Lock()
+			dis.syncedChan <- struct{}{}
+			dis.synced = true
+			close(dis.syncedChan)
+			dis.mu.Unlock()
+		})
+
+		_ = wait.PollUntilContextCancel(informerCtx, time.Second, true, func(ctx context.Context) (bool, error) {
+			if sharedIndexInf.HasSynced() {
+				syncOnce()
+				return true, nil
+			}
+			return false, nil
+		})
+	}()
+
+	return nil
+}
+
+func (dis *dynamicInformerSource) WaitForSync(ctx context.Context) error {
+	dis.mu.Lock()
+	// if we haven't already started, dis.ctx (the informer context) is not populated
+	// and will cause a panic
+	if !dis.started {
```

Comment:
Food for thought, similar to `dis.synced`, we might be able to have a `startedChan` that we close to signal that the informer has started. If we do that, can we maybe completely remove the mutex? 

Do we need a `sync.Once` for `close(startChan)` or are we okay with panicing on closing a closed channel?